src/Microsoft.Diagnostics.Monitoring/ReversedServerConnectionsSource.cs

```diff
@@ -0,0 +1,194 @@
+            {
+                try
+                {
+                    ReversedDiagnosticsConnection connection = await _server.AcceptAsync(token);
+
+                    await _connectionsSemaphore.WaitAsync(token);
+
+                    _connections.Add(connection);
+
+                    _connectionsSemaphore.Release();
+
+                    if (null != _newConnectionSource)
+                    {
+                        _newConnectionSource.TrySetResult(null);
+                    }
+                }
+                catch (OperationCanceledException)
+                {
+                }
+            }
+        }
+
+        /// <summary>
+        /// Tests a diagnostics connection and removes it if the associated runtime instance is no longer active.
+        /// </summary>
+        private async Task PruneConnectionAsync(ReversedDiagnosticsConnection connection, CancellationToken token)
+        {
+            // Cancel event pipe processing within reasonable amount of time to remove potentially unresponsive processes.
+            using CancellationTokenSource timeoutSource = new CancellationTokenSource(TimeSpan.FromSeconds(5));
+
+            // Try to get ProcessInfo information to test the connection
+            try
+            {
+                DiagnosticsClient client = new DiagnosticsClient(connection.Endpoint);
+
+                // NOTE: This test does not need to necessarily use the event pipe providers and collection.
+                // It only needs to send some type of request and receive a response from the runtime instance
+                // to verify that the diagnostics pipe/socket is viable and the runtime instance is still connected.
+                await using DiagnosticsEventPipeProcessor process = new DiagnosticsEventPipeProcessor(PipeMode.ProcessInfo);
+
+                using CancellationTokenSource linkedSource = CancellationTokenSource.CreateLinkedTokenSource(token, timeoutSource.Token);
+
+                await process.Process(client, 0, Timeout.InfiniteTimeSpan, linkedSource.Token);
+            }
+            catch (OperationCanceledException) when (token.IsCancellationRequested)
+            {
+                // Do not remove connection if method was cancelled.
+                throw;
+            }
+            catch (Exception)
+            {
+                // Runtime instance likely no longer exists or is not responsive; remove connection.
+                _connections.Remove(connection);
+
+                // Dispose the connection to release the tracking resources in the reversed server.
+                connection.Dispose();
+            }
+        }
+
+        /// <summary>
+        /// Waits for a new connection to be made available to the connection source from the reversed server.
+        /// </summary>
+        /// <param name="token">The token to monitor for cancellation requests.</param>
+        internal async Task WaitForNewConnectionAsync(CancellationToken token)
```

Comment:
Is this meant for testing only? If there's some desire to observe that a new connection has occurred, perhaps an event can be raised, rather than having an external caller wedge a task in the middle of processing. This also seems very dangerous. What if this method is called multiple times? The second task will override the first completely.

Reply:
IAsyncEnumerable can also come in handy here. You can advance the iteration each time a new connection is made.